### PR TITLE
[WIP] Fix mypy typing for `utilities.auto_restart`

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?

Part of #7037. This PR disentangles large #7218. It fixes mypy typing in `utilities.auto_restart.py`

There's one last tough mypy issue:
```
torch_lightning/utilities/auto_restart.py:86: error: Argument 1 to "len" has incompatible type "Union[Sampler[Any], Generator[Any, Any, Any], Iterator[Any]]"; expected "Sized"  [arg-type]
```

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -116,12 +116,13 @@ def _compute_current_iteration(self, num_batches_processed: Optional[int] = None
...
         return current_iteration
 
-    def _load_cached_state(self):
+    def _load_cached_state(self) -> None:
         if self._cached_state_dict is None or self.worker_id not in self._cached_state_dict:
             return
         self._current_iteration = self._cached_state_dict[self.worker_id]["current_iteration"]
         # delete cached state, prevent reloading every time iter() is called
         self._cached_state_dict = None
+        return
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:125]: pretty sure this is not necessary :/

File: pytorch_lightning/utilities/auto_restart.py [hunk 2]
```diff
@@ -143,8 +144,10 @@ def __init__(self, dataset: IterableDataset) -> None:
     def sampler(self) -> Sampler:
         return self.dataset.sampler
 
-    def state_dict(self) -> Dict[str, Any]:
-        return {k: v.state_dict() for k, v in self.samplers.items()}
+    def state_dict(self) -> Optional[Dict[str, Any]]:
+        if isinstance(self.samplers, dict):
+            return {k: v.state_dict() for k, v in self.samplers.items()}
+        return None
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:150]: I think this should assert that self.samplers is a dict or return an empty dict, not None

File: pytorch_lightning/utilities/auto_restart.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:125]: should we put `warn_no_return = False` for mypy?

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -328,22 +332,26 @@ def _cycle_to_next_worker_and_reset(dataloader: DataLoader, state_dict: Dict[str
...
+            out[iterator_k] = iterator_v
 
     if not isinstance(dataloader.dataset, CaptureIterableDataset):
         fast_forward_sampler = _find_fast_forward_samplers(dataloader)
         if fast_forward_sampler is not None:
-            out.update(fast_forward_sampler.state_dict(num_batches_processed=num_batches_processed))
+            for sampler_k, sampler_v in fast_forward_sampler.state_dict(
+                num_batches_processed=num_batches_processed
+            ).items():
+                out[sampler_k] = sampler_v
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:350]:
```suggestion
            samplers_state_dict = fast_forward_sampler.state_dict(num_batches_processed=num_batches_processed).items()
            for sampler_k, sampler_v in samplers_state_dict.items():
                out[sampler_k] = sampler_v
```

File: pytorch_lightning/utilities/auto_restart.py (hunk 1 above)
Comment (stancld) [reply pytorch_lightning/utilities/auto_restart.py:125]: I remember I read a PEP recommendation suggesting if there's any `return` statement within a function, it should be obeyed for the whole function definition. But yes, my suggestion might be redundant here.
